### ISSUE TITLE
http server and server connnector to expose restful api

### ISSUE THREAD

**Issue by pi0** [unjs/db0#6]
Similar to [unstorage server](https://unstorage.unjs.io/http-server) this would allow exposing sql backends as a restful API to edge runtime 🐦

**Comment by amandesai01** [comment]
Is this up for grabs? I am open to implement it.

**Comment by pi0** [comment]
Surely @amandesai01 feel free to open a draft if you like! (i can continue on top of your work then in same PR)

**Comment by amandesai01** [comment]
Can I get some initial spec (need not be detailed)

Basically what will be input vs output?

**Comment by amandesai01** [comment]
For now, this is what I believe is expected:
One endpoint to take SQL query and prepared statements, and return result as JSON.

What else?

**Comment by atinux** [comment]
I think this would be quite similar to https://starbasedb.hashnode.space/default-guide/rest-api/introduction

Checkout their implementation on https://github.com/Brayden/starbasedb/blob/main/src/literest/index.ts

**Comment by pi0** [comment]
Yep, but more i'm thinking to expose remote RPC as raw/unprepared queries rather than CRDT verbs. Any higher over layer usage can then extend server to build something similar but their own parsing and conventions.

**Comment by amandesai01** [comment]
Can you show some sample implementation / links? @pi0

**Comment by pi0** [comment]
It is a new idea/API here so I don't have an example to share.. But I will try to draft something as soon as had little time to move it forward :)

**Comment by sandros94** [comment]
> It is a new idea/API here so I don't have an example to share.. But I will try to draft something as soon as had little time to move it forward :)

I theory we could also have multiple implementation right? I'm still not fully sure if I understand what I'm reading around, but something like a `crosswire` (much like `crossws` and `y-crossws`) based on [Supabase's `pg-gateway`](https://github.com/supabase-community/pg-gateway) should be possible right?

**Comment by alex-eri** [comment]
> * POST endpoint

`... RETURNING id` is wide used for insert, `... RETURNING *` is good to see how defauls set.

**Comment by amandesai01** [comment]
A couple of proposals were made for this.

Proposal 1:
https://github.com/unjs/db0/issues/6#issuecomment-2437598835

Proposal 2:
https://github.com/unjs/db0/issues/6#issuecomment-2519488700

Proposal 3:
https://github.com/unjs/db0/issues/6#issuecomment-2525678643

What should the actual implementation be closest to?

**Comment by pi0** [comment]
Lets go with the simplest HTTP spec (https://github.com/unjs/db0/issues/6#issuecomment-2437598835), we an later build SSE (and possibly MCP) atop 🚀

**Comment by amandesai01** [comment]
Or maybe we can have a mix of 1 and 2:

1. GET and POST as per https://github.com/unjs/db0/issues/6#issuecomment-2437598835
2. And then POST on `/raw` as per https://starbasedb.hashnode.space/default-guide/http-endpoints/query

**Comment by pi0** [comment]
Feel free to make PR, we can iterate and discuss 😊

**Comment by amandesai01** [comment]
Hi, apologies for the delay. I needed a server for a quick project I have been working on. Built following based on agreed spec 1 (happy to contribute it):

Goes without saying that following code is an MVP and doesn't do a lot of stuff like setting the right headers and authentication.

This has been working well for me.

```ts
import { defineHandler, readBody, getQuery, createError, HTTPError } from "h3";
import type { Database } from "db0";

export interface RawHandlerOptions {
  db: Database;
}

type Param = string | number | boolean | undefined | null;

function throwBadRequest(message: string): never {
  throw new HTTPError(message, { status: 400 });
}

function normalizeParams(params: Param | Param[]): Param[] {
  if (!params) return [];
  return Array.isArray(params) ? params : [params];
}

/**
 * Creates a raw SQL handler that executes queries directly
 *
 * - GET: Executes query with db.prepare().all() and returns rows
 * - POST: Executes statement with db.prepare().run() and returns { success: boolean }
 */
export function createHandler(options: RawHandlerOptions) {
  const { db } = options;

  const handler = defineHandler(async (event) => {
    const method = event.req.method;

    if (method === "GET") {
      const query = getQuery<{ sql?: string; params?: string | string[] }>(
        event,
      );
      if (!query.sql) throwBadRequest("SQL query is required");
      const statement = db.prepare(query.sql);
      return { rows: await statement.all(...normalizeParams(query.params)) };
    }

    if (method === "POST") {
      const body = await readBody<{ sql?: string; params?: string | string[] }>(
        event,
      );
      if (!body?.sql) throwBadRequest("SQL query is required");
      const statement = db.prepare(body.sql);
      return {
        success: (await statement.run(...normalizeParams(body.params))).success,
      };
    }

    throw new HTTPError("Method not allowrd", { status: 405 });
  });

  return handler;
}

```
